lib/Service/SessionService.php

```diff
@@ -0,0 +1,101 @@
+ * @copyright Copyright (c) 2022, chandi Langecker (git@chandi.it)
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Deck\Service;
+
+use OCA\Deck\Db\Session;
+use OCA\Deck\Db\SessionMapper;
+use OCA\Deck\Event\SessionCreatedEvent;
+use OCA\Deck\Event\SessionClosedEvent;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\EventDispatcher\IEventDispatcher;
+use OCA\NotifyPush\Queue\IQueue;
+use OCP\Security\ISecureRandom;
+
+class SessionService {
+	public const SESSION_VALID_TIME = 92;
+
+	private SessionMapper $sessionMapper;
+	private ITimeFactory $timeFactory;
+	private $userId;
+	private IEventDispatcher $eventDispatcher;
+	private ISecureRandom $secureRandom;
+
+	public function __construct(
+		SessionMapper $sessionMapper,
+		ISecureRandom $secureRandom,
+		ITimeFactory $timeFactory,
+		$userId,
+		IEventDispatcher $eventDispatcher
+	) {
+		$this->sessionMapper = $sessionMapper;
+		$this->secureRandom = $secureRandom;
+		$this->timeFactory = $timeFactory;
+		$this->userId = $userId;
+		$this->eventDispatcher = $eventDispatcher;
+	}
+
+	public function initSession($boardId): Session {
+		$session = new Session();
+		$session->setBoardId($boardId);
+		$session->setUserId($this->userId);
+		$session->setToken($this->secureRandom->generate(64));
+		$session->setLastContact($this->timeFactory->getTime());
+
+		$session = $this->sessionMapper->insert($session);
+		$this->eventDispatcher->dispatchTyped(new SessionCreatedEvent($boardId, $this->userId));
+		return $session;
+	}
+
+	/**
+	 * @throws DoesNotExistException
+	 */
+	public function syncSession(int $boardId, string $token) {
+		$session = $this->sessionMapper->find($boardId, $this->userId, $token);
+		$session->setLastContact($this->timeFactory->getTime());
+		$this->sessionMapper->update($session);
+	}
+
+	public function closeSession(int $boardId, string $token): void {
+		try {
+			$session = $this->sessionMapper->find($boardId, $this->userId, $token);
+			$this->sessionMapper->delete($session);
+		} catch (DoesNotExistException $e) {
+		}
+		$this->eventDispatcher->dispatchTyped(new SessionClosedEvent($boardId, $this->userId));
+	}
+
+	public function notifyAllSessions(IQueue $queue, int $boardId, $event, $excludeUserId, $body) {
```

Comment:
Just because I was curious how things behave, I added a small patch that would also notify on content changes (which is the next step but seems to work quite nicely already). In case you need that as it is kind of the next step:

```diff
--- a/lib/Db/ChangeHelper.php
+++ b/lib/Db/ChangeHelper.php
@@ -23,10 +23,14 @@

 namespace OCA\Deck\Db;

+use OCA\Deck\NotifyPushEvents;
+use OCA\Deck\Service\SessionService;
+use OCA\NotifyPush\Queue\IQueue;
 use OCP\ICacheFactory;
 use OCP\ICache;
 use OCP\IDBConnection;
 use OCP\IRequest;
+use OCP\Server;

 class ChangeHelper {
        public const TYPE_BOARD = 'boardChanged';
@@ -49,12 +53,22 @@ class ChangeHelper {
                $this->userId = $userId;
        }

-       public function boardChanged($boardId) {
+       public function boardChanged(int $boardId) {
                $time = time();
                $etag = md5($time . microtime());
                $this->cache->set(self::TYPE_BOARD . '-' . $boardId, $etag);
                $sql = 'UPDATE `*PREFIX*deck_boards` SET `last_modified` = ? WHERE `id` = ?';
                $this->db->executeUpdate($sql, [$time, $boardId]);
+               try {
+                       $queue = \OCP\Server::get(IQueue::class);
+                       \OCP\Server::get(SessionService::class)
+                               ->notifyAllSessions($queue, $boardId, NotifyPushEvents::DeckBoardUpdate, $this->userId, [
+                                       'id' => $boardId
+                               ]);
+               } catch (\Exception $e) {
+                       // most likely notify_push is not installed.
+                       return;
+               }
        }

        public function cardChanged($cardId, $updateCard = true) {
```

Reply:
probably good to be aware about what comes next :)

So far I've created a listener ([Application.php](https://github.com/nextcloud/deck/pull/3876/files#diff-4eb824426a8cffb8b33b62b57744c69f32ea30ea755249a7a6170d9402885deb) -> [LiveUpdateListener.php](https://github.com/nextcloud/deck/pull/3876/files#diff-9037a36df135b63c0274adf1d2b46b28b51fd1f16067c8bc1acdbef4b1a23df6)) which is listening for OCP events and forwarding them then to the related sessions.

This seems cleaner for me , because I don't need to hook into all the individual methods for changing things and it's a more standardized way than the `ChangeHelper` methods.

@juliushaertl or would you prefer the `ChangeHelper` way?